### ISSUE TITLE
Is Rust struct FFI on the table?

### ISSUE THREAD

**Issue by Zireael07** [dpbriggs/x7#1]
From the blog, I understand this is not yet the case...

**Comment by dpbriggs** [comment]
Hi! It's not very stable but it is in use in [redis-oxide](https://github.com/dpbriggs/redis-oxide/blob/40ca1809418106d1fdc9fd610bd755026121788b/src/scripting.rs#L159). There's an [FFI example](https://github.com/dpbriggs/x7/blob/b07963102c08914e6a6831c04f912392db89c70c/examples/ffi.rs#L1) as well:

```
➜  x7 git:(master) ✗ cargo run --example=ffi
   Compiling x7 v0.2.0 (/home/david/programming/x7)
    Finished dev [unoptimized + debuginfo] target(s) in 1.39s
     Running `target/debug/examples/ffi`
(mydata-sum 1 1)                 -> Ok(Int(2))
(mydata-sum "the number is: " 1) -> Ok(String("the number is: 1"))
(fn (x) x)                       -> Err(MyError("Cannot convert Fn<AnonFn, 1, [ x ]> to MyData!"))
```

I wouldn't consider any of this stable. I've been profiling x7 and have been making breaking changes as a result.

As for "Rust Struct" - do you mean some way to convert a whole rust struct into like an x7 record? Something like this maybe?

```rust

struct Foobar {
    a: u32
    b: u32
}

// Not sure what the API should be? A proc-macro maybe?
let foo = Foobar { a: 3, b: 4 }
X7Interpreter::add_record(foo)

// (.a foo) ;; 3
```

**Comment by Zireael07** [comment]
Yes, exactly what I mean. I have Rust structs that I want to have available on the scripting side - most likely as records in x7 parlance.

**Comment by dpbriggs** [comment]
Hmmm that'll be interesting to add. We'll want a convenient API and that'll take some design work. What sort of application are you hoping to embed x7 into?

A workaround for now could be to let the objects live on the rust side and add x7 functions to interact with them. I use that [approach](https://github.com/dpbriggs/redis-oxide/blob/40ca1809418106d1fdc9fd610bd755026121788b/src/scripting.rs#L209) in redis-oxide. The trick is to move the item you want to interact with into the closure (an `Arc<Mutex<T>>` could work here nicely)
